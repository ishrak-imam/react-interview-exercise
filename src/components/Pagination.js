import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import styles from './Pagination.css'

class Pagination extends Component {
  constructor (props) {
    super(props)
    this.calculatePages = this.calculatePages.bind(this)
    this.goToPage = this.goToPage.bind(this)
  }

  goToPage (page) {
    return () => {
      if (this.props.currentPage !== page && !(page < 1) && !(page > this.props.pages.length)) {
        this.props.goToPage(page)
      }
    }
  }

  calculatePages (pages, page) {
    const currentPage = Number(page)

    if (pages.length <= 4) return pages
    const pagesChunk = []

    if (currentPage === pages.length) {
      pagesChunk.push('dot')
      for (let i = currentPage - 2; i <= currentPage; i++) {
        pagesChunk.push(String(i))
      }
      return pagesChunk
    }

    if (currentPage === 1) {
      for (let i = 1; i <= currentPage + 2; i++) {
        pagesChunk.push(String(i))
      }
      pagesChunk.push('dot')
      return pagesChunk
    }

    if (pages.length > 4 && Math.ceil(pages.length / 2) > currentPage) {
      for (let i = currentPage - 1; i <= currentPage + 1; i++) {
        pagesChunk.push(String(i))
      }
      pagesChunk.push('dot')
    }
    if (pages.length > 4 && Math.ceil(pages.length / 2) <= currentPage) {
      pagesChunk.push('dot')
      for (let i = currentPage - 1; i <= currentPage + 1; i++) {
        pagesChunk.push(String(i))
      }
    }
    return pagesChunk
  }

  render () {
    let { pages, currentPage } = this.props

    const chunkedPages = this.calculatePages(pages, currentPage)

    return (
      <div className={styles.pagination}>
        <button onClick={this.goToPage('1')} className={`${styles.anchor}`}>&laquo;</button>
        <button onClick={this.goToPage(String(Number(currentPage) - 1))} className={`${styles.anchor}`}>&lsaquo;</button>
        {chunkedPages.map(page => {
          if (page === 'dot') return <span className={`${styles.anchor}`} key={page}>. . .</span>
          return (
            <button
              onClick={this.goToPage(page)}
              key={page}
              className={classnames(styles.anchor, {
                'active': currentPage === page
              })}
            >{page}</button>
          )
        })}
        <button onClick={this.goToPage(String(Number(currentPage) + 1))} className={`${styles.anchor}`}>&rsaquo;</button>
        <button onClick={this.goToPage(String(pages.length))} className={`${styles.anchor}`}>&raquo;</button>
      </div>
    )
  }
}

Pagination.propTypes = {
  pages: PropTypes.array.isRequired,
  currentPage: PropTypes.string.isRequired
}

export default Pagination
