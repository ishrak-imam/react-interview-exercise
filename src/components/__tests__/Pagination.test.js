import React from 'react'
import { create } from 'react-test-renderer'
import Pagination from '../Pagination'

test('Pagination component', () => {
  const component = create(
    <Pagination pages={[]} currentPage='' />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
