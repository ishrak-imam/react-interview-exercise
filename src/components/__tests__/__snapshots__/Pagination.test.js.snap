exports[`test Pagination component 1`] = `
<div
  className={undefined}>
  <button
    className="undefined"
    onClick={[Function]}>
    «
  </button>
  <button
    className="undefined"
    onClick={[Function]}>
    ‹
  </button>
  <button
    className="undefined"
    onClick={[Function]}>
    ›
  </button>
  <button
    className="undefined"
    onClick={[Function]}>
    »
  </button>
</div>
`;
